src/main/java/com/wavefront/generators/TraceGenerator.java

```diff
@@ -76,28 +88,47 @@ protected void startGeneration(boolean isRealTime, @Nonnull GeneratorConfig gene
         // Simulate the delay.
         current += SLEEP_DELAY_MILLIS;
       }
-      mustBeGeneratedSpans = (long) (rate * (current - start) / 1000);
+      mustBeGeneratedSpans = (rate * (current - start) / 1000);
 
       while (generatedSpans < mustBeGeneratedSpans) {
         final Trace trace = generateTrace(current - RANDOM.nextInt(SLEEP_DELAY_MILLIS));
         if (trace != null) {
           dataQueue.addTrace(trace);
           generatedSpans += trace.getSpansCount();
-        }else {
+        } else {
           break;
         }
+        updateHeapMemory();
+        sleeping(isRealTime);
       }
+    }
+    logger.info("Generation complete!\n" + ANSI_YELLOW + String.format(generatorConfig.getGeneratorConfigFile() +
+        " Memory " + "usage- %.2fGB / %.2fGB", usedHeapMemoryGB, maxHeapMemoryGB) + ANSI_RESET);
+    sendStat(str);
+  }
 
-      if (isRealTime) {
-        try {
-          Thread.sleep(SLEEP_DELAY_SECONDS);
-        } catch (InterruptedException e) {
-          logger.severe(Throwables.getStackTraceAsString(e));
-        }
+  private void sleeping(boolean isRealTime) {
+    while (isRealTime && flag) {
+      if (dataQueue.size() <= (maxDataQueueSize * QUEUE_PERCENTAGE) / 100) {
+        flag = false;
+      }
+      try {
+        Thread.sleep(SLEEP_DELAY_MILLISECONDS);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
       }
     }
-    sendStat(str);
-    logger.info("Generation complete!");
+  }
+
+  private void updateHeapMemory() {
+    double currentUsed = (double) memoryMXBean.getHeapMemoryUsage().getUsed() / GIGA;
+    if (usedHeapMemoryGB < currentUsed) {
+      usedHeapMemoryGB = currentUsed;
+    }
+    if (currentUsed >= (maxHeapMemoryGB * MEMORY_PERCENTAGE) / 100) {
```

Comment:
`(maxHeapMemoryGB * MEMORY_PERCENTAGE) / 100` always has the same value. Maybe it makes sense to calculate it only once?